kafka: add time_offset option

uce opened the pull request:
@elindsey I was unsure who to ping for review, so I picked you, because you had a look at the (internal) spec linked in #6689 before and have experience with Kafka sources. Please let me know whether I should pick somebody else.

As discussed in the (internal) spec, we will probably move offsets and timestamps to columns and allow efficient filtering which is more general and idiomatic than the WITH options we currently have.

@frankmcsherry also raised a point about the name `time_offset`. Should we rename it to `kafka_time_offset` to avoid any future confusion? I'm onboard with it, but wanted to keep the discussion in the PR. I'll update the docs after some input on this question. Edit: renamed to `kafka_time_offset`.

uce commented on src/sql/src/kafka_util.rs:
```diff
@@ -246,6 +250,132 @@ pub async fn test_config(
+///   before now (e.g. `-10` means 10 millis ago)
+///
+/// If `time_offset` has not been configured, an empty Option is returned.
+pub async fn lookup_start_offsets(
```
We could merge this with the existing `test_config` to avoid another `fetch_metadata` call, but I left it as is to avoid premature optimization and keep the concerns separate.

uce replied:
Thanks! I'm reusing the client from the `test_config` fn now (renamed to `create_consumer` which still seems reasonable for the existing doc comment).

uce commented on src/sql/src/kafka_util.rs:
```diff
@@ -246,6 +250,132 @@ pub async fn test_config(
+            }
+        };
+
+    task::spawn_blocking({
```
I was unsure whether we prefer multiple smaller scoped calls to `spawn_blocking` for the individual requests or just have one.

JLDLaughlin replied:
I'm also unsure! The only other call to `spawn_blocking` that I see is in [`test_config`](https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/kafka_util.rs#L215), which doesn't answer our question. @petrosagg or @benesch would probably have the most useful advice here.

benesch replied:
My model is that every call to `spawn_blocking` has approximately as much overhead as creating a new thread. You can think of `spawn_blocking` as equivalent to:

```rust
let (tx, rx) = std::sync::mpsc::channel();
thread::spawn({
    let result = /* do stuff */;
    tx.send(result);
});
rx.recv();
```

So that's an argument for the general rule being "use fewer calls to `spawn_blocking`", but also feel free to make your own determination based on the situation.

petrosagg replied:
I didn't know the answer to this so I looked at the implementation and ran a small experiment. I think it's ok to have many **sequencial** calls to `spawn_blocking`. Tokio spawns blocking threads on demand but crucially, it keeps them around for some time (10 seconds by default) and reuses them for subsequent blocking needs.

This test program spawns just two extra threads:

```rust
use std::{thread, time::Duration};

#[tokio::main]
async fn main() {
    println!("runtime started");
    for _ in 0..100 {
        let _ = tokio::task::spawn_blocking(|| {
            thread::sleep(Duration::from_millis(10));
        }).await;
    }
    println!("blocking done");
    thread::sleep(Duration::from_millis(10_010));
    println!("workers should be gone now");
}
```

`strace` output:

```rust
strace -f -e trace=clone ./target/debug/tokio-block-test
clone(child_stack=0x7f47c35fedb0, [...]) = 338157
strace: Process 338157 attached
[pid 338156] clone(child_stack=0x7f47c33fddb0, [..]) = 338158
strace: Process 338158 attached
[pid 338156] clone(child_stack=0x7f47c31f9db0, [..]) = 338159
strace: Process 338159 attached
[pid 338156] clone(child_stack=0x7f47c2ff5db0, [..]) = 338160
strace: Process 338160 attached
runtime started
[pid 338156] clone(child_stack=0x7f47c2df4db0, [..]) = 338161
strace: Process 338161 attached
[pid 338156] clone(child_stack=0x7f47c2beadb0, [..]) = 338162
strace: Process 338162 attached
blocking done
[pid 338161] +++ exited with 0 +++
[pid 338162] +++ exited with 0 +++
workers should be gone now
[pid 338157] +++ exited with 0 +++
[pid 338158] +++ exited with 0 +++
[pid 338159] +++ exited with 0 +++
[pid 338160] +++ exited with 0 +++
+++ exited with 0 +++
```

uce replied:
Thank you very much for your inputs!  I'll stick to the single `spawn_blocking` call here since the blocking API calls are very close to each other and not much happens in between calls. Still, good to know that the threads would likely be re-used and the mental model is very helpful, too. 🚀

elindsey commented on src/sql/src/kafka_util.rs:
```diff
@@ -246,6 +250,132 @@ pub async fn test_config(
+            let mut time_offsets = elems.iter();
+            let mut start_offsets = Vec::with_capacity(num_partitions as usize);
+
+            for pid in 0..num_partitions {
+                let offset = match time_offsets.next() {
+                    Some(partition) => match partition.offset() {
+                        Offset::Offset(offset) => offset,
+                        _ => fetch_end_offset(&consumer, &topic, pid)?,
+                    },
+                    _ => fetch_end_offset(&consumer, &topic, pid)?,
+                };
+                start_offsets.push(offset);
+            }
```
In what cases does Kafka return results for fewer partitions than we asked for, is that an expected case we need to gracefully handle? If it's only obscure cases like a partition resize as the query is being purified, I'd be inclined to simplify this and treat that as an invariant violation, eg. (crappy rust pseudocode)

```
let start_offsets = offsets_for_times.iter().map({
    if let Offset::offset(o) {
        o
    } else {
        fetch_end_offset
    }}).collect();

if start_offsets.len() != num_partitions {
  // throw user-facing invariant violation error
}
```

uce replied:
The case is independent of partition resizing and occurs if the provided timestamp value is greater than all timestamps for a partition (e.g. skew between partitions or a timestamp far "in the future"). My gut feeling is that it wouldn't be a too common case, but I also think it doesn't hurt to handle it.

I'm not too opinionated here, so feel free to insist on simplifying the code. If we keep it, I'll add a comment to clarify why we do it.

elindsey replied:
This may be me totally misunderstanding the underlying API calls, but I'll take a stab at articulating what I'm concerned about to see if it's legit. 🙂 

My main concern isn't about coming up with end offsets for ts that are out of range, but that iteration is based on num_partitions and we currently match on the None branch of `time_offsets.next()`, so may synthesize offsets for pids not in the response.

---

We create an API request for ListOffsets with a payload like: "pid 0 offset ts FOO, pid 1 offset ts FOO, pid 2 offset ts FOO"

My expectation is that we get back a result for each pid we requested, though the individual partition result may be Offset::Invalid or some form of error to represent FOO was an out of range ts.
eg. response might be: "pid 0 offset 1, pid 1 offset invalid/error, pid 2 offset 3"
We do want to map those invalid offset results to the tail offset of the topic if the ts is out of range, as you mentioned.

What I would be surprised to see is a response with a subset of pids we asked for, such as:
"pid 0 offset 1, pid 1 offset 1"
or worse
"pid 1 offset 1, pid 2 offset 2"

If we never see responses of that form, then the code is correct but the match on None/current iteration style is unnecessary and could hide problems. If we do see responses of that form then the code may be wrong because we're relying on the implicit pid ordering but the response is potentially sparse (we never validate `TopicPartitionListElem.partition()` but instead infer it).

I think my questions are:
1. Are we guaranteed that all pids in our request show up in our response?
2. Are we guaranteed the ordering of pids in the response?

(my hunch is that the answer to both is yes)

uce replied:
That's also how I understood your first message. Sorry for being unclear!

The answer to both questions is Yes. Originally, I was under the wrong impression that the answer to 1. is No, because I only checked it against `Offset::Invalid` whereas the response is `Offset::End` (which makes sense).

I'll update the PR.

uce replied:
Just pushed a fix in https://github.com/MaterializeInc/materialize/pull/6941/commits/f195b30f7cdaad6d0e866d044a5e2b9461f1f703.

I left it as a separate commit for easier reviewing.

elindsey reviewed (commented):
> Please let me know whether I should pick somebody else.

I did a quick pass and left a couple of Kafka questions, but I'll defer to @JLDLaughlin for review - she has a deeper understanding of the sql crate and previously worked on `start_offsets`.

You'll also want to add docs/release notes:

Look at `start_offset` here as an example/to see where to add this new param in the docs
doc/user/layouts/partials/create-source/connector/kafka/with-options.html
doc/user/layouts/partials/create-source/connector/kafka/details.html

You can add it to release notes here in the v0.7.4 section
doc/user/content/release-notes.md

JLDLaughlin commented on src/sql/src/kafka_util.rs:
```diff
@@ -11,14 +11,18 @@
 
 use std::collections::BTreeMap;
 use std::convert;
+use std::convert::TryInto;
```
tiny nit: I think we would usually try to write these imports like `use std::convert::{self, TryInto};`

uce replied:
Thanks and done! I _thought_ these were merged automatically by `rustfmt`, but I apparently was confused. 🙈

JLDLaughlin replied:
`rustfmt` will reorder your imports! but it doesn't fix these things, unfortunately. (I've learned the hard way! 😂 )

JLDLaughlin commented on src/sql/src/kafka_util.rs:
```diff
@@ -11,14 +11,18 @@
 use log::{debug, error, info, warn};
 use rdkafka::client::ClientContext;
 use rdkafka::consumer::{BaseConsumer, Consumer, ConsumerContext};
+use rdkafka::Offset;
```
you could do something similar here, if you'd like! (`use rdkafka::{Offset, TopicPartitionList}`)

uce replied:
Thanks and done! I _thought_ these were merged automatically by `rustfmt`, but I apparently was confused. 🙈

frankmcsherry replied:
There is an option, but we may have it disabled as it is quite assertive. 

https://rust-lang.github.io/rustfmt/?version=v1.4.36&search=#imports_granularity

JLDLaughlin commented on src/sql/src/kafka_util.rs:
```diff
@@ -246,6 +250,132 @@ pub async fn test_config(
+            let num_partitions: i32 = consumer
+                .fetch_metadata(Some(&topic), Duration::from_secs(10))?
+                .topics()
+                .first()
```
It looks like we'll want to check if `fetch_metadata` returns more than one result -- we do that in the [`sink_connector` code](https://github.com/MaterializeInc/materialize/blob/main/src/coord/src/sink_connector.rs#L72). I think you could also pull `get_partitions` into `kafka_util`, if you're interested!

uce replied:
Thanks for the hint! Move to `kafka_util`.

JLDLaughlin commented on src/sql/src/pure.rs:
```diff
@@ -93,6 +93,38 @@ pub fn purify(
                     config_options = kafka_util::extract_config(&mut with_options_map)?;
                     kafka_util::test_config(&broker, &topic, &config_options).await?;
+
+                    // Translate `time_offset` to `start_offset`.
```
👍🏻

JLDLaughlin reviewed (commented):
the sql bits look good to me!

uce commented:
Thanks for the input, everyone! I updated the PR to address all comments and I've also renamed the option to `kafka_time_offset` as mention in the original description.

@elindsey There is one remaining question regarding [handling of fewer offsets than partitions](https://github.com/MaterializeInc/materialize/pull/6941#discussion_r645740795). I'm inclined to keep the code as is (and add a comment), but also open to simplify it as you suggest. What do you think?

uce commented on src/sql/src/kafka_util.rs:
```diff
@@ -243,12 +250,142 @@ pub async fn test_config(
+                    Offset::End => fetch_end_offset(&consumer, &topic, pid as i32),
+                    _ => bail!("Unexpected offset {:?} for partition", elem.offset()),
+                })
+                .collect::<Result<Vec<i64>, anyhow::Error>>()?;
```
Regarding the type annotation: The borrow checker complains that it can't figure out the type which might be a hint that I'm doing something weird. Happy to restructure to be more idiomatic.

uce replied:
Thanks! I changed it to infer the contained types. 👍🏼

elindsey reviewed (approved):
lgtm, thank you, that's a lot easier to follow!

I left a few small things/style notes, but your discretion if any those are worth changing.

uce commented on src/sql/src/kafka_util.rs:
```diff
@@ -243,12 +250,142 @@ pub async fn test_config(
+                .map(|(pid, elem)| match elem.offset() {
+                    Offset::Offset(offset) => Ok(offset),
+                    Offset::End => fetch_end_offset(&consumer, &topic, pid as i32),
+                    _ => bail!("Unexpected offset {:?} for partition", elem.offset()),
```
Whops! Makes sense, done!

uce commented on src/sql/src/kafka_util.rs:
```diff
@@ -243,12 +250,142 @@ pub async fn test_config(
+            let start_offsets = offsets_for_times
+                .elements()
+                .iter()
+                .enumerate()
```
👍🏼 Done!
